ingestion-edge/ingestion_edge/disk_queue.py

```diff
@@ -0,0 +1,70 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, you can obtain one at http://mozilla.org/MPL/2.0/.
+
+from .conf import QUEUE_FILE
+from .conf import QUEUE_POSITION_FILE
+from flask import Blueprint
+from google.cloud.pubsub import PublisherClient
+from os import remove
+from os.path import exists
+from threading import Lock
+from typing import Tuple
+import ujson
+
+disk_queue = Blueprint("disk_queue", __name__)
+client = PublisherClient()
+read_lock = Lock()
+write_lock = Lock()
+
+
+def write(**kwargs):
+    with write_lock:
+        # TODO disable autoscaling
+        with open(QUEUE_FILE, "a") as q:
+            q.write(ujson.dumps(kwargs) + "\n")
+            q.flush()
+
+
+@disk_queue.route("/__flush__")
+def flush_queue() -> Tuple[str, int]:
+    if not exists(QUEUE_FILE):
+        # early return for empty queue
+        return "", 204
+    with read_lock:
+        # initialize queue state
+        done = False
+        start = 0
+        end = 0
+        if exists(QUEUE_POSITION_FILE):
+            with open(QUEUE_POSITION_FILE, "r") as qp:
+                start = int(qp.read())
+        try:
+            # process the queue
+            # TODO release write lock between line reads
+            with write_lock:
+                # TODO rotate queue file and handle rotated files first
+                with open(QUEUE_FILE, "r") as q:
+                    for line in q:
+                        if end < start:
+                            # TODO use q.seek(start)
+                            end += 1
```

Comment:
My assumption on first read-through was that `start` and `end` were byte positions in the file, but looks like they're line numbers? Will that need to be refactored in order to use `seek`? We might want to use more specific variable names to make this easier to follow.

Reply:
I agree that using byte positions would be better.

Reply:
yes. the `TODO` comment here is for using byte positions,